### ISSUE TITLE
Create different type of feeds based on mathematical models

### ISSUE THREAD

**Issue by mlodic** [intelowlproject/GreedyBear#36]
Right now, there are 2 feeds available based on the `age` of the indicator:

* recent: most recent IOCs seen in the last 3 days
* persistent: these IOCs are the ones that were seen regularly by the honeypots. This feeds will start empty once no prior data was collected and will become bigger over time.

But they are really simple. Maybe we could think about creating different kind of lists based on more reliable values.
This issue will need a complex research analysis.
Starting point could be the Stratosphere's research: https://www.stratosphereips.org/aip-tool

**Comment by regulartim** [comment]
I would like to work on this as part of my Master's thesis. Is there a possibility to talk to you about the topic?

**Comment by mlodic** [comment]
sure, can you contact me in private in the GSOC Honeynet Slack or Twitter?

* [Slack chat](https://gsoc-slack.honeynet.org/)
* [Twitter](https://twitter.com/intel_owl)

**Comment by regulartim** [comment]
The PR #392 closed itself when I renamed the source branch. I will open another PR when I'm ready. Sorry for that.

**Comment by mlodic** [comment]
np

**Comment by regulartim** [comment]
After testing several models I have some results. The best performing ones are Random Forest models. Here is a short comparison between them and the age filters already implemented in GreedyBear:

| Model | Interaction Recall | IP F1 Score | IP Precision | IP Recall |
|---|---|---|---|---|
| Recent (GreedyBear) | 0.5490 | 0.4675 | 0.4212 | 0.5253 |
| Persistent (GreedyBear) | 0.3871 | 0.6063 | 0.5462 | 0.6812 |
| Random Forest Classifier | 0.4532 | **0.6263** | **0.5642** | **0.7037** |
| Random Forest Regressor | **0.8315** | 0.1629 | 0.1468 | 0.1831 |

In simple terms the recall scores describe what fraction of interactions / IPs would have been blocked today if I had used yesterday's feed as a blocklist (assuming a feed size of 5000 IOCs). 

As you can see, the Random Forest Regressor is optimised for predicting the number of interactions that can be expected from an IP address and is clearly superior to all other models in this respect.

The Random Forest Classifier performs only slightly better than the persistent feed. However, it is more flexible as it can be of any size, whereas the persistent feed is limited by the number of IOCs that meet its criteria. 

What do you think? Which of them should we integrate into GreedyBear?

**Comment by mlodic** [comment]
Wonderful analysis! :) 

> What do you think? Which of them should we integrate into GreedyBear?

I think that they both make sense as possible options and they are not exclusive to each other or with the already existing "algorithms". Users can then choose between which list to extract

So feel free to integrate both of them

**Comment by regulartim** [comment]
Thank you! :) Both models produce some sort of score. I'd like to run the models once a day and calculate the score for all the IOCs and then store it in the DB. Then we can create feeds by ordering by the scores.

**Comment by regulartim** [comment]
I have a third model in the works that focuses on the threat level of an IOC. I will also integrate that one.

**Comment by regulartim** [comment]
I guess I am ready to create these feeds now, but I struggle to find good names for them. I am thinking of "most expected hits" and "likely to recur". What do you think?

Here are some suggestions on how to integrate them:
1. Just add them next to "recent" and "persistent" as a different "age" class (which is a little misleading)
2. Add them next to "recent" and "persistent" but rename "age" to "prioritize".
3. Replace  "recent" and "persistent" and rename "age" to "prioritize".
4. Something else.

**Comment by mlodic** [comment]
I am open to any of these possibilities but if I have to choose I would pick the first one cause that would simplify the integration. "likely_to_recur" is easy to understand at a high level and normal users don't have to bother about the magic behind. I like it in that sense. Then, obviously, there's a lot of stuff behind. And I think that this value is better found in the documentation for advanced users.
That's the approach that I have used with discrete success in IntelOwl. On one hand, give easy stuff to use and understand to the people then, on the other hand, provide all the knowledge to understand what is really happening and how to leverage it for expert users.
Does it make sense to you?

**Comment by regulartim** [comment]
Yes it does. If that is fine for you, I will implement my 2nd suggestion.

**Comment by mlodic** [comment]
perfect! :)

**Comment by regulartim** [comment]
After the decision to not merge #459 I consider this completed.
